### ISSUE TITLE
Model: Door Locks

### ISSUE THREAD

**Issue by garrynewman** [Facepunch/rust#99]
We need to look into how the locks work on the current doors. They look like shit and they don't fit.

The key lock doesn't have a model, and the code lock is too high res, the LODs are awful and it doesn't fit anywhere properly on the doors.

I think it would be best if Vince came up with the lock models so it stays consistent with the current style.

**Comment by paroxum** [comment]
I can have a look at this after the barricades are done. Do we want each door type to have its own lock or just a global redesign of the current lock?

**Comment by garrynewman** [comment]
Ideally one model should fit all. The lock position can change per door though.

**Comment by PaulBradley87** [comment]
Just a thought i always had that any kind of keypad mechanism for the door, should be located on the wall to the right of the door handle rather then on the door itself, might make it easier too so you dont have to worry about how it interacts with any different crazy doors if you might wanted in future (like say if you wanted a raising or sliding bunker door type thing).

**Comment by garrynewman** [comment]
Yeah - I noticed that in one of howie's concepts and it did look a lot better, definitely something to look into.

**Comment by PaulBradley87** [comment]
![rustdoorlock](https://cloud.githubusercontent.com/assets/11909909/7113379/0c112f98-e1cb-11e4-82c2-7f98c58721e1.jpg)
 so this is how i kinda did the door lock in a paintover i did a few weeks ago, something a bit simpler and easy to read from far away

**Comment by paroxum** [comment]
I like the idea an electric wire running along for a while before going through the wall. In terms of the look, do we want to keep the openned box with wires and batteries look or should we rather make it look homebrew but closed. My personal opinion is that a closed design might read better from a distance as you can better control what the noise level is on the piece.

**Comment by paroxum** [comment]
The electric cable running down the wall may not be very realistic though. As you'd imagine you can bypass the box with wire cutters here. Up to how real we want to be.

**Comment by PaulBradley87** [comment]
yea, i can understand that for the wire. i defo agree tho that sometimes you can overdesign stuff... this i think can be a case of just keep it more simple, give it some cool details that you see up close that indicate its handmade, and give it a good read for players to see and i think itll just look alot less awkward then it does now.

**Comment by vegeta897** [comment]
One would assume the system would default to lockdown if the wires were cut. Or if the keypad were bashed in with a rock. Or shot. Whether the wire is exposed doesn't make much of a difference in the possibility of being destroyed somehow.

**Comment by scotthdavidson** [comment]
I would be hesitant to add a wire due to the plan on having a whole electricity system.

**Comment by paroxum** [comment]
good point about the electricity system. I wont add the wire as it seems to add more visual confusion with how electricity might work in future.

**Comment by PaulBradley87** [comment]
cant we just have an inside element to the lock, maybe have some wiring on the inside? anything to add some cool little details the walls is a good thing

**Comment by scotthdavidson** [comment]
I think an inside element is fine. I would just avoid any wires to nowhere as once electricity comes in will will need to be stripped out if the system uses wires.

**Comment by garrynewman** [comment]
Yeah I wouldn't add the wire at the bottom. It probably wouldn't work on the different material walls anyway.

Not married to the exposed circuit boards. I think the less detail the better really - so it's readable from normal standing position. The problem with the old one is that it only really looks good in screenshots.

**Comment by PaulBradley87** [comment]
![doorcodelockpositionideas](https://cloud.githubusercontent.com/assets/11909909/7134359/138f12fc-e294-11e4-9354-39b72655ed0a.jpg)
![doorcodelockideassimplified](https://cloud.githubusercontent.com/assets/11909909/7134363/18ea570c-e294-11e4-97f0-fa791956f67c.jpg)
did some rough ideas for a simpler keypad lock with a more solid design if you wanna try that. they look crappy in black and white but it kinda gives the idea.

**Comment by scotthdavidson** [comment]
Add some big old brackets to hold it to the wall and I think most of them will work.

**Comment by paroxum** [comment]
sure scott. I have a couple ideas to form something out of 2,4 and 7. I'll post a greybox here when I have one to see if everybody likes it.

**Comment by scotthdavidson** [comment]
My thinking is that as we move up though each tier of locks the screen gets bigger so you can tell each type apart.

**Comment by paroxum** [comment]
![keypad_greybox_01](https://cloud.githubusercontent.com/assets/11387355/7155860/dd735838-e359-11e4-8330-a664cdd5fc94.jpg)
 Here is a first pass on the greybox, I think we can extract two variations from one bake on this, in order to make lower tier versions for 'free', using the same texture. It's a bit of a mix from the concepts, and I might make things less boxy for the HP but still into this silhouette.

**Comment by scotthdavidson** [comment]
What's your thinking is the gameplay difference between these 3?

**Comment by paroxum** [comment]
At this stage this is just visual differences, I would assume they all play the same role. Unless you put a hacking minigame in where it gets harder with every tier :P

The attempt at making silhouette slightly different is just to make it easy to spot from a distance.

**Comment by paroxum** [comment]
![image](https://cloud.githubusercontent.com/assets/11387355/7156080/b69ea364-e35b-11e4-9348-e40275c960ec.png)
Did another version where I pushed the scale difference further here

**Comment by scotthdavidson** [comment]
I think the idea is that each Tier of lock has a different functions. For example:

Tier1 - Standard lock
Tier2 - Combination lock (Like you get on a suitcase)
Tier3 - Keypad lock
Tier4 -Palm Print lock
Tier5 - Retinal lock

Each type of lock would have it's benefits, drawbacks. each would have it's own unique way of cracking.

**Comment by scotthdavidson** [comment]
Personally I don't think a hacking minigame has enough going on.

For example, If you want to get past a Palm print lock there are multiple options. 

Find a guy who lives in the house, kill him and chop his hand off.
Craft a specific tool for the job.

plus all the other existing ways of doing it.

**Comment by paroxum** [comment]
Is that where we're headed though? Because from an art POV having these different components means more work obviously, right now if they all operate the same we can make variation easily.

Will we have the gameplay to back it up if we make 5 tiers that all operate differently?

**Comment by garrynewman** [comment]
Changing the lock based on the building type isn't something we want to do. It's not technically attached to the building, it's attached to the door. So in code it knows nothing about the building. We can attach by assuming a wall is there (because you can only place a door in a doorway).

I don't think it's a good idea to automatically specialize the lock design depending on the door grade either. If anything they should be different locks and the user should choose which one they want to place. It works from a design point of view - but we'd be robbing the user of choice from a gameplay pov.

**Comment by paroxum** [comment]
Alright, so we're headed toward different types of lock then. Can we figure out how many types we need then in order to have an idea of the work? 5 seems like a lot to me.

**Comment by scotthdavidson** [comment]
It's also worth the cost of art for a new lock type if it adds loads of new player choice. I would just focus on the first 3 locks for now.

**Comment by paroxum** [comment]
Tier1 - Standard lock: what do we talk about for this one?
Tier2 - Combination lock: Personally I think instead of the suitcase, something more like a vault with knobs here?
Tier3 - Keypad lock

**Comment by scotthdavidson** [comment]
Howie did a bunch of standard locks/handles designs a while ago but I think it depends on if the standard lock has to fit on the door or can sit next to the door like the others.

**Comment by paroxum** [comment]
yeah that's why I wondered, also, combination lock: http://www.rarityguide.com/articles/content_images/1/downpour/safepuzzledownpour.jpg

**Comment by garrynewman** [comment]
I would just stick with a one key lock and one codelock right now. We can expand from there.

**Comment by paroxum** [comment]
Okay ;-) I'm going to rework the greybox a bit more then, logical order for me doesnt change, will make the keypad one first, and can follow up with the standard padlock.

**Comment by Helkus** [comment]
@paroxum those keypad locks don't really look like they belong in rust, they look a bit like they are from battlefield 2142 or bladerunner let me gather some examples and I'll post them there

**Comment by paroxum** [comment]
Helk, that's not the direction I'm taking.
![keypad_detail_greybox_01](https://cloud.githubusercontent.com/assets/11387355/7160898/df789362-e383-11e4-888a-899eec4533e9.jpg)
Here is a detailed greybox to show a more precise idea of what I have in mind in terms of style/era/silhouette. This should help seeing where this is going better :)

**Comment by paroxum** [comment]
the gas mask like component in the middle will be the speaker, if anybody wonders :P

**Comment by peturfp** [comment]
I have to agree with Helk. It's a good design, but it doesn't really look like something you'd find on an improvised shack.

It looks like something manufactured by Weyland Yutani.

**Comment by Helkus** [comment]
I'm not sure how to say this but when I see that the last game on earth I would think of is Rust.
Fallout, Halo, maybe the Jetsons come to mind instead

It's too large, the angles are way too 1960's and it just doesn't fit

**Comment by Helkus** [comment]
I've asked Paul to do some more concepts so why don't we reconvene when those are complete, sorry dude :x

**Comment by paroxum** [comment]
We might be running into something then concerning keypads here. Whatever they look like they are going to end up looking out of place on low tier walls.

That's alright if it's not everybody's taste in the end. I personally think it woudl fit the top tier segment of the game perfectly visually.

**Comment by PaulBradley87** [comment]
yea i agree, i think this an awesome little asset, but i think it just needs to be really simple to fit with current buildings. The challenge with rust crafted stuff is that when you push something to look cool it sorta suddenly becomes a product/industrialy produced, but when you make it super hand made and functional it just becomes flat and boring, its a tricky balance. im gonna paint up something similar to 1,6 and 8 from my thumbs because the design is so simple its sorta hard to get the feel until some texture and grime is thrown in there, really it wont look too far from howies origional concept but without the exposed circuit stuff.

**Comment by Helkus** [comment]
the real problem with current keypad is the model is very unoptimized, the texture too large and terrible LOD's and you can tell the texture wasn't painted by the same person who did the doors

**Comment by paroxum** [comment]
Yeah for this kind of object it's hard to get something that looks better than the generic keypad we have now without putting some art direction into it. I can see why this issue arises. We'll see what Paul comes with in his next concept. I might finish the other one in my freetime as a portfolio piece. :)

**Comment by garrynewman** [comment]
It doesn't have to be a keypad, It's just a combination lock. So it doesn't need a screen or any kind of electricity.

On another note. The different tiers of keypads could have fewer numbers. So the first tier would have just 2 buttons.. which is still kind of secure but at the same time is easier to guess the combination. The buttons don't have to be numeric, could shapes, letters, colours etc.

**Comment by Utilitron** [comment]
Mechanical Code lock, no electricity required.
http://www.americanlocksets.com/kaba-simplex-7100-series-7102-pushbutton-lock-234-backset-p-4602.html
http://simplexlock.com/pdfs/products/69/7102/7100%20series%20parts.pdf

**Comment by scotthdavidson** [comment]
![image](https://cloud.githubusercontent.com/assets/11922686/7163424/d76527ae-e391-11e4-95c1-f2e55415a0ea.png)

1-4 tumblers, chunky and  clear.

**Comment by y0himba** [comment]
I am not an artist, I cannot draw it, but try and picture this: A wooden faceplate with 5 small wood levers sticking out.  

They must be depressed in a player set order to push up small piston type wood pegs inside that will fit into groves in a pattern to release the lock on the door. 

This will be primitive looking and still allow for a combination, a player set pattern. 

The levers can be labeled with numbers, characters or letters, and this would also allow the possibility of the order being guessed, in essence 'hacking' the code.

It is entirely plausible that something like this can be carved from a blueprint. The lock can use the existing "keypad" that we use now for entering a numerical code, or a new one can be created.

From a player point of view, this fits, and offers a new way to loot a cabin or a base.

**Comment by y0himba** [comment]
Frank Walls was good enough to create some mockups for me:

Inside there are plungers.  When depressed in the correct order, they allow the door to open. it is plausible.

![pua7txh](https://cloud.githubusercontent.com/assets/11462603/7176208/4924511e-e3ea-11e4-83c9-17a1187cb32a.jpg)

![nvp2lru](https://cloud.githubusercontent.com/assets/11462603/7176215/4c5a0946-e3ea-11e4-99c1-79f92758bfc6.jpg)

**Comment by PaulBradley87** [comment]
still workin on a few more of these, but as this one is done thoi8ght id throw it up to show you the direction i had in mind
![rustcodelock1](https://cloud.githubusercontent.com/assets/11909909/7182498/af6223de-e449-11e4-9831-b559b003f593.jpg)

**Comment by garrynewman** [comment]
That might work without the inner bevels, if it was just one faceplate. Can you try some without the numbers - just with a red/green light showing open or locked.

**Comment by PaulBradley87** [comment]
sure thing

**Comment by PaulBradley87** [comment]
![rustcodelock2](https://cloud.githubusercontent.com/assets/11909909/7182803/9fcb46b0-e44b-11e4-88ed-11bc70438901.jpg)
this is how it'd look

**Comment by Helkus** [comment]
pretty good! something I don't like about how rounded/beveled the edges of the internal panel are tho, can we try one with smaller bevels or square, just to compare?

**Comment by garrynewman** [comment]
From an engineering point of view, if I was making that, would it have the round inside bevel or would it be one flat piece. Would the light be the size of the panel, or would it just be a single bulb.

**Comment by PaulBradley87** [comment]
Here another idea going super simple like you said Garry, no led just basic bulbs and a single metal plate housing the lights and keypad
![rustcodelock3](https://cloud.githubusercontent.com/assets/11909909/7185027/5c392374-e458-11e4-8ad5-b6c348cb56fd.jpg)

**Comment by garrynewman** [comment]
It's definitely getting closer. the lights feel a bit big to me, but that might be its charm. I'd move the keyboard to the side and have a bunch of holes for where the sound comes out. The hard part is going to be making it look right in game - I'm thinking in the end we'll have two have it twice it's actual size to be usable.

**Comment by peturfp** [comment]
Maybe exaggerate the box surface area, but keep the keys realistic in size.

**Comment by peturfp** [comment]
P.p.s. Would it be cool to have it thick enough to clip through the entire wall, so that you'd see the back of it inside hte house, with maybe cables going down into the floor?

A free interior decoration, of sorts.

**Comment by PaulBradley87** [comment]
hah yea - i just went ingame and realised it has to make a sound, ill play about with it some more with that in mind.

**Comment by frank-walls** [comment]
Here's a couple versions. These only use 9 keys though.
![rust_keypad2](https://cloud.githubusercontent.com/assets/11983304/7188550/683b8598-e414-11e4-9cc7-22c67e5fc0a6.jpg)
![rust_keypad1](https://cloud.githubusercontent.com/assets/11983304/7188551/683dab7a-e414-11e4-9b21-665bf822fde1.jpg)

**Comment by garrynewman** [comment]
I think we need to think more about what it's going to look like in situ. A blown up drawing of the panel doesn't help when it's going to be about 20x20 pixels on screen.

**Comment by Helkus** [comment]
![image](https://cloud.githubusercontent.com/assets/4874057/7191680/760765ea-e45d-11e4-8198-122d1d0cbfc5.png)

just had a play

**Comment by garrynewman** [comment]
try making it 2x bigger and sticking it on the wall to the right of the door

**Comment by Helkus** [comment]
2x is welcome to doom

![image](https://cloud.githubusercontent.com/assets/4874057/7191797/44eb6d52-e45e-11e4-85c2-739b89c8c3be.png)


1.5x
![image](https://cloud.githubusercontent.com/assets/4874057/7191833/6c7c9954-e45e-11e4-95c4-bd5e667e3c71.png)




1.25x

![image](https://cloud.githubusercontent.com/assets/4874057/7191849/89af70f0-e45e-11e4-9523-46b1022cbac5.png)

**Comment by Helkus** [comment]
2x welcome to doom
![image](https://cloud.githubusercontent.com/assets/4874057/7191877/d55f9c8c-e45e-11e4-80c9-6e1d7b3fa8fb.png)

1.5x
![image](https://cloud.githubusercontent.com/assets/4874057/7191879/daee08e6-e45e-11e4-9dc4-a33a07bd35e4.png)


1.25x
![image](https://cloud.githubusercontent.com/assets/4874057/7191882/dfa33aa0-e45e-11e4-9f28-296f05f61a92.png)

**Comment by garrynewman** [comment]
Haha 2x looks like the face scanners from HL1. 1.5 is probably the limit where it would get weird afterwards.

**Comment by frank-walls** [comment]
Sorry to keep hijacking your workflow, lol. Can't help it - you should probably close this to the public. Here's a pixel art keypad.
![rust_keypad_tiny](https://cloud.githubusercontent.com/assets/11983304/7192240/79ba8484-e42f-11e4-81ff-cbb46a204421.jpg)

**Comment by Helkus** [comment]
wow, here's a pixel art banana
![image](https://cloud.githubusercontent.com/assets/4874057/7192325/71748ff8-e462-11e4-9be4-9516f8f0e146.png)

**Comment by thoresh** [comment]
Omg helk, so good. When can we expect this?

**Comment by SgtFlexxx** [comment]
I think it would look better if it was just a small bulb that protrudes from the panel that lights up for a few seconds red or blue depending on of the code entered is wrong or right. Also should possibly include the '0' key for gameplay purposes

**Comment by SgtFlexxx** [comment]
Red or green* my bad

**Comment by frank-walls** [comment]
![rust_banana](https://cloud.githubusercontent.com/assets/11983304/7192645/161fde0c-e433-11e4-8111-75db322bfd85.jpg)

**Comment by OverlordJ19** [comment]
These are all super great. However I think there should definitely be some way to hack in, it is no fun waiting 3 hours for an afk player to come back and open his door. Then you still have to kill him before he shuts the door.

**Comment by PaulBradley87** [comment]
Heres a rough idea what it could look like by the door 
![Uploading Codelockconceptmockups.jpg . . .]()
with a speaker built in too.

**Comment by AlexanderLaheij** [comment]
Those look good. I have a slightly crazy request, which is probably ahead of its time. The request is that if you enter a code wrong 2 a 3 times it will trigger - a trap . Brainstorm on that :+1: please ;)

**Comment by garrynewman** [comment]
That's working for me.

@paroxum when modelling this and placing it next to the wall, do you think we're going to have problems with the different wall widths? (It needs to be double ended, so it pokes through the other side too.

**Comment by paroxum** [comment]
yes we will.

**Comment by paroxum** [comment]
Although we can try it and see if it works, end result might just pop out more on thinner walls.

**Comment by garrynewman** [comment]
I guess it needs to just pop out slightly more than the door frame to be safe?

We should open a new issue for the key lock - because that hasn't got a model at all right now.

**Comment by paroxum** [comment]
If everyone is happy with the last concept I'll give that a go next week.

**Comment by Helkus** [comment]
I like the first and last, of that most recent concept 

one thing to keep in mind about this entity is people deploy them on other things too like storage crates

**Comment by ChazFales** [comment]
Can't wait this is going to be awesome. where does the power come from for the lock? maybe use the batteries that are in the game. Give the lock a battery life like the decay, batteries will run out after three days or so. just a thought looking great. one more thing if you have a speaker on the lock can we mute the players in side unless standing right next to it. thanks
